sql/binlog.cc

```diff
@@ -8282,6 +8454,14 @@ bool MYSQL_BIN_LOG::write_cache(THD *thd, binlog_cache_data *cache_data,
     update_thd_next_event_pos(thd);
   }
 
+#ifdef HAVE_REPLICATION
+  if (binlog_space_limit && binlog_space_total &&
+      binlog_space_total + my_b_tell(&log_file) > binlog_space_limit)
+  {
+    purge_logs_by_size(true);
```

Comment:
I am not sure we should purge potentially after each write. Purging after close/reopen/rotate etc. operations - i.e. at file granularity - should be OK, even if that means that the space limit will be intermittently violated by a larger value. Stalls in a middle of binlog files would be unexpected to users.

Reply:
`max_binlog_size` doesn't wait for close/reopen/rotate. When log is bigger than `max_binlog_size`, server rotates and purges logs. Doc says:
```
If a write to the binary log causes the current log file size to exceed the value of this variable, the server rotates the binary logs
```

Reply:
I designed `binlog_space_limit` after `relay_log_space_limit` which is also checked in each `next_event()` call.

Reply:
OK then. Just make sure that no filesystem binlog enumeration happens for each binlog write, but only in the event of space limit violation.

Reply:
`binlog_space_total` is created to reduce binlog enumeration. It is updated after log rotation or RESET MASTER. The following condition:
```
  if (binlog_space_limit && binlog_space_total &&
      binlog_space_total + my_b_tell(&log_file) > binlog_space_limit)
```
is true only when logs need to be purged. After purging logs the condition will be false.
In other words `purge_logs_by_size(true)` should be called only once per each purged file.